cmd/kaspawallet/daemon/server/split_transaction.go

```diff
@@ -0,0 +1,194 @@
+}
+
+func (s *server) mergeTransaction(splitTransactions []*serialization.PartiallySignedTransaction,
+	originalTransaction *serialization.PartiallySignedTransaction, splitWalletAddress *walletAddress) (
+	*serialization.PartiallySignedTransaction, error) {
+
+	targetAddress, err := util.NewAddressScriptHash(originalTransaction.Tx.Outputs[0].ScriptPublicKey.Script, s.params.Prefix)
+	if err != nil {
+		return nil, err
+	}
+	changeAddress, err := util.NewAddressScriptHash(originalTransaction.Tx.Outputs[1].ScriptPublicKey.Script, s.params.Prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	totalValue := uint64(0)
+	sentValue := originalTransaction.Tx.Outputs[0].Value
+	utxos := make([]*libkaspawallet.UTXO, len(splitTransactions))
+	for i, splitTransaction := range splitTransactions {
+		output := splitTransaction.Tx.Outputs[0]
+		utxos[i] = &libkaspawallet.UTXO{
+			Outpoint: &externalapi.DomainOutpoint{
+				TransactionID: *consensushashing.TransactionID(splitTransaction.Tx),
+				Index:         0,
+			},
+			UTXOEntry:      utxo.NewUTXOEntry(output.Value, output.ScriptPublicKey, false, constants.UnacceptedDAAScore),
+			DerivationPath: s.walletAddressPath(splitWalletAddress),
+		}
+		totalValue += output.Value
+		totalValue -= feePerInput
+	}
+
+	var payments []*libkaspawallet.Payment
+	if totalValue >= sentValue {
+		payments = []*libkaspawallet.Payment{{
+			Address: targetAddress,
+			Amount:  sentValue,
+		}, {
+			Address: changeAddress,
+			Amount:  totalValue - sentValue,
+		}}
+	} else {
+		// sometimes the fees from compound transactions make the total output higher than what's available from selected
+		// utxos, in such cases, the remaining fee will be deduced from the resulting amount
+		payments = []*libkaspawallet.Payment{{
+			Address: targetAddress,
+			Amount:  totalValue,
+		}}
+	}
+	mergeTransactionBytes, err := libkaspawallet.CreateUnsignedTransaction(s.keysFile.ExtendedPublicKeys,
+		s.keysFile.MinimumSignatures, payments, utxos)
+	if err != nil {
+		return nil, err
+	}
+
+	return serialization.DeserializePartiallySignedTransaction(mergeTransactionBytes)
+}
+
+func (s *server) maybeSplitTransaction(transaction *serialization.PartiallySignedTransaction,
+	splitAddress util.Address) ([]*serialization.PartiallySignedTransaction, error) {
+
+	transactionMass, err := s.estimateMassAfterSignatures(transaction)
+	if err != nil {
+		return nil, err
+	}
+
+	if transactionMass < mempool.MaximumStandardTransactionMass {
+		return []*serialization.PartiallySignedTransaction{transaction}, nil
+	}
+
+	splitCount := int(transactionMass / mempool.MaximumStandardTransactionMass)
+	if transactionMass%mempool.MaximumStandardTransactionMass > 0 {
+		splitCount++
+	}
+	inputCountPerSplit := len(transaction.Tx.Inputs) / splitCount
+
+	splitTransactions := make([]*serialization.PartiallySignedTransaction, splitCount)
+	for i := 0; i < splitCount; i++ {
+		startIndex := i * inputCountPerSplit
+		endIndex := startIndex + inputCountPerSplit
```

Comment:
```
	inputCountPerSplit := len(transaction.Tx.Inputs) / splitCount
	if len(transaction.Tx.Inputs)%splitCount > 0 {
		splitCount++
	}
```
You missed this part (note the logic is a bit weird, I'm incrementing `splitCount`, and not `inputCountPerSplit`, since incrementing `inputCountPerSplit` might make mass too high.
I'll add a comment to clarify that.